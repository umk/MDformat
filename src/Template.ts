import { Placeholder, Ref, TemplateParts } from './TemplatePart'

const ID_REGEX = '[_\\p{L}][_\\p{L}\\p{Nd}]*'

/**
 * A regular expression that matches a sequence of JavaScript
 * property names separated by dots. This supports only a subset of
 * the JavaScript property names.
 */
const PATH_REGEX = `${ID_REGEX}(\\.${ID_REGEX})*`

/**
 * A regular expression that matches template data field reference. The
 * path denotes the path through object hierarchy to the data field, that
 * substitutes the reference. The group is optional and denotes a key,
 * that is used to zip the data field values with the same group value.
 */
const REF_REGEX = new RegExp(`\\{\\{((?<group>${ID_REGEX}):)?(?<path>${PATH_REGEX})\\}\\}`, 'gu')

type Template = {
  /**
   * The template parts.
   */
  parts: TemplateParts
  /**
   * A collection of fields referenced from the template
   * @internal
   */
  refs?: Array<Ref>
} & { __brand: 'template' }

export function getTemplateRefs(template: Template): Array<Ref> {
  let refs = template.refs
  if (!refs) {
    template.refs = refs = template.parts.filter((p) => typeof p === 'object') as Array<Ref>
  }
  return refs
}

export function createTemplate(template: string): Template {
  const matches = Array.from(template.matchAll(REF_REGEX)).map((match) => ({
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    index: match.index!,
    raw: match[0],
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    path: match.groups!.path,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    group: match.groups!.group ?? match.groups!.path,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    modifier: match.groups!.modifier as string | undefined,
  }))
  const parts: TemplateParts = []
  // const refs: Map<string, Ref> = new Map()
  const refs: Record<string, Record<string, Ref>> = {}
  let lastIndex = 0
  for (const { index, raw, path, group } of matches) {
    parts.push(template.slice(lastIndex, index))
    lastIndex = index + raw.length
    const properties = path.split('.')
    const ref: Placeholder = { path, properties, group, raw }
    ;(refs[group] || (refs[group] = {}))[path] = ref
    parts.push(ref)
  }
  parts.push(template.slice(lastIndex))
  return {
    parts,
    refs: Object.values(refs).flatMap((g) => Object.values(g)),
  } as Template
}

export default Template
