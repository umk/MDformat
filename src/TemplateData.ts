import { configuration } from './Configuration'
import Content, { ContentPrimitive } from './Content'
import Template, { getTemplateRefs } from './Template'
import { Ref } from './TemplatePart'

/**
 * A mapping from a group name to an object that contains
 * another mapping from a data reference field to substitution
 * for that field.
 */
type TemplateData = {
  groups: Record<string, TemplateDataGroup>
}

export type TemplateDataGroup = {
  /**
   * Index of the record in the group taken as a substitution
   * for placeholder.
   */
  index: number
  refs: TemplateDataRefs
}

/**
 * A mapping from a data reference field to an object that
 * contains substitution for that field. The key is provided
 * in exact wording that provided in the template.
 */
export type TemplateDataRefs = Record<string, Array<Content>>

export function getTemplateData(
  templateOrRefs: Template | Array<Ref>,
  content: Content,
): Array<TemplateData> {
  const refs = Array.isArray(templateOrRefs) ? templateOrRefs : getTemplateRefs(templateOrRefs)
  const refsByProperty = refs.reduce(
    (refByProperty, { path, properties: [property, ...rest], group }) => {
      let refGroup = refByProperty.get(property)
      if (!refGroup) {
        refByProperty.set(property, (refGroup = []))
      }
      refGroup.push({
        path,
        properties: rest,
        group,
      })
      return refByProperty
    },
    new Map<string | undefined, Array<Ref>>(),
  )
  const refsArr = Array.from(refsByProperty.entries())
  const items = resolveContent(content)
  return items.flatMap((item, index, values) =>
    Array.from(refsArr)
      .map(([property, samePropRefs]): Array<TemplateData> => {
        // If there is a property to traverse in order to reach
        // the placeholder value...
        if (property) {
          // ...if property value is an object...
          if (typeof item === 'object' && item !== null && property in item) {
            // ...get the substitution value from that property
            const data: Content = item[property]
            return getTemplateData(samePropRefs, data)
          } else {
            // ...otherwise substitution could not be found
            return [{ groups: {} }]
          }
        } else {
          // ...otherwise use current value as a placeholder substitution
          const onePropResult = samePropRefs.reduce<TemplateData>(
            (previous, { group, path }) => {
              let dataGroup = previous.groups[group]
              if (!dataGroup) {
                previous.groups[group] = dataGroup = { index, refs: {} }
              }
              dataGroup.refs[path] = values
              return previous
            },
            {
              groups: {},
            },
          )
          return [onePropResult]
        }
      })
      .sort((a, b) => {
        // Sort the results by the number of items for greater
        // possibility to shrink the number of results based on
        // groupings or missing properties.
        return a.length - b.length
      })
      .reduce(permutateData, [{ groups: {} }]),
  )
}

function permutateData(a: Array<TemplateData>, b: Array<TemplateData>): Array<TemplateData> {
  const result: Array<TemplateData> = []
  for (const itemA of a) {
    itemsB: for (const itemB of b) {
      const item: TemplateData = { groups: {} }
      const entries = [...Object.entries(itemA.groups), ...Object.entries(itemB.groups)]
      for (const [group, { index, refs }] of entries) {
        const groupData = item.groups[group]
        if (!groupData) {
          item.groups[group] = { index, refs: { ...refs } }
        } else {
          if (groupData.index !== index) {
            // Being in the same group, two properties have
            // different indexes and cannot be zipped.
            continue itemsB
          }
          item.groups[group] = { index, refs: { ...groupData.refs, ...refs } }
        }
      }
      result.push(item)
      if (result.length > configuration.maxPermutations) {
        throw new Error('The template resolves into too many combinations')
      }
    }
  }
  return result
}

function resolveContent(content: Content): Array<ContentPrimitive | Record<string, Content>> {
  if (Array.isArray(content)) {
    return content.flatMap((c) => resolveContent(c))
  }
  return [content]
}

export default TemplateData
