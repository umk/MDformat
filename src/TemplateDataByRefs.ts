import Content from './Content'
import TemplateData from './TemplateData'
import { Ref } from './TemplatePart'

type TemplateDataByRefs = {
  /**
   * The template data for grouped refs.
   */
  key: TemplateData
  /**
   * A collection of template data outside of the grouping.
   */
  data: Array<TemplateData>
}

export function groupDataByRefs(
  data: Array<TemplateData>,
  refs: Array<Ref>,
): Array<TemplateDataByRefs> {
  // Group refs by their group ID
  const refGroups: Record<string, Array<Ref>> = {}
  for (const ref of refs) {
    ;(refGroups[ref.group] || (refGroups[ref.group] = [])).push(ref)
  }
  return Object.entries(refGroups).reduce<Array<TemplateDataByRefs>>(
    (prev, [groupId, refs]) =>
      prev.flatMap(({ key, data }) => {
        // The template data records are allocated into buckets
        // by an index of referenced item in current group, then...
        const groups = data
          .reduce<Array<Array<TemplateData>>>((prev, cur) => {
            const group = cur.groups[groupId]
            if (group) {
              const { index } = group
              ;(prev[index] || (prev[index] = [])).push(cur)
            }
            return prev
          }, [])
          .map(
            (data, index): TemplateDataByRefs => ({
              key: {
                groups: {
                  ...key.groups,
                  [groupId]: { index, refs: {} },
                },
              },
              data,
            }),
          )
        // ...allocate buckets themselves into smaller buckets by
        // an array value, that contains substitution.
        return refs.reduce(
          (prev, ref) =>
            prev.flatMap(({ key, data }) => {
              const dataByRef = data.reduce((prev, cur) => {
                const value = cur.groups[groupId].refs[ref.path]
                let arr = prev.get(value)
                if (!arr) prev.set(value, (arr = []))
                arr.push(cur)
                return prev
              }, new Map<Array<Content>, Array<TemplateData>>())
              return Array.from(dataByRef.entries()).map(([value, data]): TemplateDataByRefs => {
                const currentGr = key.groups[groupId]
                return {
                  key: {
                    groups: {
                      ...key.groups,
                      [groupId]: {
                        ...currentGr,
                        refs: {
                          ...currentGr.refs,
                          [ref.path]: value,
                        },
                      },
                    },
                  },
                  data,
                }
              })
            }),
          groups,
        )
      }),
    [{ key: { groups: {} }, data }],
  )
}

export default TemplateDataByRefs
