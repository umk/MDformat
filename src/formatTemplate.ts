import Template from './Template'
import TemplateData from './TemplateData'
import { isPlaceholder } from './TemplatePart'

/**
 * Creates a content that represents the template bound to the data
 * object, where each of the data object properties can be referenced
 * from the template.
 * @param template The template structure.
 * @param data A template data record
 * @returns A string that represents the template bound to the data object.
 */
function formatTemplate(template: Template, data: TemplateData): string {
  return template.parts
    .map((part) => {
      if (!isPlaceholder(part)) {
        return part
      }
      const { path, group } = part
      const d = data.groups[group]
      if (d) {
        const { index, refs } = data.groups[group]
        const ref = refs[path]
        if (ref) {
          let content = ref[index]
          if (part.formatter) {
            content = part.formatter(content)
          }
          return content
        }
      }
      return part.raw
    })
    .join('')
}

export default formatTemplate
