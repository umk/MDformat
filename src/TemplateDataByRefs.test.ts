import { createTemplate, getTemplateRefs } from './Template'
import { getTemplateData } from './TemplateData'
import { groupDataByRefs } from './TemplateDataByRefs'
import formatTemplate from './formatTemplate'

describe('groupDataByRefs', () => {
  it('groups data by zero refs', () => {
    const template = createTemplate(
      'Lorem ipsum {{gr:dolor}} sit {{gr:amet}}, consectetur {{adipiscing}} elit',
    )
    const data = getTemplateData(template, [
      {
        dolor: ['perspiciatis', 'unde'],
        amet: ['omnis', 'iste', 'natus'],
        adipiscing: ['ex', 'ea'],
      },
      {
        dolor: ['sequi'],
        amet: ['porro', 'quisquam'],
        adipiscing: ['non', 'proident'],
      },
    ])
    const groups = groupDataByRefs(data, [])
    expect(groups).toHaveLength(1)
    expect(Object.entries(groups[0].key.groups)).toHaveLength(0)
    expect(groups[0].data).toHaveLength(6)
  })

  it('groups data by refs', () => {
    const template = createTemplate(
      'Lorem ipsum {{gr:dolor}} sit {{gr:amet}}, consectetur {{adipiscing}} elit',
    )
    const refs = getTemplateRefs(template)
    const data = getTemplateData(template, [
      {
        dolor: ['perspiciatis', 'unde'],
        amet: ['omnis', 'iste', 'natus'],
        adipiscing: ['ex', 'ea'],
      },
      {
        dolor: ['sequi'],
        amet: ['porro', 'quisquam'],
        adipiscing: ['non', 'proident'],
      },
    ])
    const groups = groupDataByRefs(
      data,
      refs.filter((r) => ['dolor', 'amet'].includes(r.path)),
    )
    const templateGr = createTemplate('{{gr:dolor}} {{gr:amet}}')
    const content = groups.map((g) => formatTemplate(templateGr, g.key)).join('\n')
    expect(content).toMatchSnapshot()
  })

  it('groups data by refs cross several groups', () => {
    const template = createTemplate(
      'Lorem ipsum {{gr:dolor}} sit {{gr:amet}}, consectetur {{adipiscing}} elit',
    )
    const refs = getTemplateRefs(template)
    const data = getTemplateData(template, [
      {
        dolor: ['perspiciatis', 'unde'],
        amet: ['omnis', 'iste', 'natus'],
        adipiscing: ['ex', 'ea'],
      },
      {
        dolor: ['sequi'],
        amet: ['porro', 'quisquam'],
        adipiscing: ['non', 'proident'],
      },
    ])
    const groups = groupDataByRefs(
      data,
      refs.filter((r) => ['dolor', 'adipiscing'].includes(r.path)),
    )
    const templateA = createTemplate('{{gr:dolor}} {{adipiscing}}')
    const templateB = createTemplate('\t{{gr:amet}}')
    const content = groups
      .map((g) => {
        const a = formatTemplate(templateA, g.key)
        const b = g.data.map((d) => formatTemplate(templateB, d))
        return [a, ...b].join('\n')
      })
      .join('\n')
    expect(content).toMatchSnapshot()
  })

  it("groups data by refs if ref data doesn't exist", () => {
    const template = createTemplate(
      'Lorem ipsum {{gr:dolor}} sit {{gr:amet}}, consectetur {{adipiscing}} elit',
    )
    const refs = getTemplateRefs(template)
    const data = getTemplateData(template, [
      {
        dolor: ['perspiciatis', 'unde'],
        // amet missing
        adipiscing: ['ex', 'ea'],
      },
      {
        dolor: ['sequi'],
        // amet missing
        adipiscing: ['non', 'proident'],
      },
    ])
    const groups = groupDataByRefs(
      data,
      refs.filter((r) => ['dolor', 'amet'].includes(r.path)),
    )
    const templateGr = createTemplate('{{gr:dolor}} {{gr:amet}}')
    const content = groups.map((g) => formatTemplate(templateGr, g.key)).join('\n')
    expect(content).toMatchSnapshot()
  })

  it('groups data by ref with and without explicit group ID', () => {
    const template = createTemplate('{{gr:testA}} {{gr:testB}} {{testA}}')
    const refs = getTemplateRefs(template)
    const data = getTemplateData(template, [
      {
        testA: ['a', 'b', 'c'],
        testB: ['1', '2', '3'],
      },
    ])
    const groups = groupDataByRefs(data, refs)
    const content = groups.map((g) => formatTemplate(template, g.key)).join('\n')
    expect(content).toMatchSnapshot()
  })
})
