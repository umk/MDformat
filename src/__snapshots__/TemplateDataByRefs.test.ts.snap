// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupDataByRefs groups data by ref with and without explicit group ID 1`] = `
"a 1 a
b 2 b
c 3 c"
`;

exports[`groupDataByRefs groups data by refs 1`] = `
"perspiciatis omnis
sequi porro
unde iste"
`;

exports[`groupDataByRefs groups data by refs cross several groups 1`] = `
"perspiciatis ex
	omnis
perspiciatis ea
	omnis
sequi non
	porro
sequi proident
	porro
unde ex
	iste
unde ea
	iste"
`;

exports[`groupDataByRefs groups data by refs if ref data doesn't exist 1`] = `
"perspiciatis {{gr:amet}}
sequi {{gr:amet}}
unde {{gr:amet}}"
`;
